docs/source/notes/autograd.rst

```diff
@@ -3,119 +3,394 @@
-every iteration. You don't have to encode all possible paths before you
-launch the training - what you run is what you differentiate.
-
-In-place operations with autograd
+    def f(w):
+        x = op_1(w)
+        y = op_2(x)
+        z = op_3(y)
+        return z
+
+    w.requires_grad_()
+    f(w).backward(v)
+
+Will change the computation to: :math:`v^T J_f = v^T J_{op_3} J_{op_2} J_{op_1} = v^T \dfrac{\partial{z}}{\partial{y}} \dfrac{\partial{y}}{\partial{x}} \dfrac{\partial{x}}{\partial{w}} = v^T \dfrac{\partial{z}}{\partial{w}}`.
+
+Handling inplace operations
+"""""""""""""""""""""""""""
+
+The only special case that needs to be considered here are inplace operations.
+The main reason to use inplace operations is to preserve side effects and prevent additional memory requirements.
+The first point means that if a tensor is referenced in different places, the inplace operation should modify all these references.
+The second point means that no extra memory allocation should be done under the hood otherwise it would defeat the purpose of inplace operations.
+
+To be able to fulfill these two needs, we make inplace operations actually be inplace during the forward pass giving exactely the same behavior
+as any other non-autograd library that provides inplace operations.
+To be able to compute the gradients, we consider that there are different versions of the tensor that existed and that the user tensor always points to the latest version.
+This allows us to write the automatic differentiation rules as before.
+
+See the two functions below for examples where we use subscripts to specify each version of the Tensor.:
+
+.. code::
+
+    def f(x):
+        z = op_1(x)
+        # z points to z_0
+        z.op_2_()
+        # z points to z_1
+        return z
+
+    def g(x):
+        z = op_1(x)
+        # z points to z_0
+        y = z.select(0, 0)
+        # y is a view into z
+        # y points to y_0 = z_0.select(0, 0)
+        z.op_2_()
+        # both z and y now points to y_1 = z_1.select(0, 0)
+        return y
+
+The corresponding autograd computations can simply be writen using the implicit variables: :math:`v^T J_f = v^T J_{op_2\_} J_{op_1} = v^T \dfrac{\partial{z_1}}{\partial{z_0}} \dfrac{\partial{z_0}}{\partial{x}} = v^T \dfrac{\partial{z_1}}{\partial{x}} = v^T \dfrac{\partial{z}}{\partial{x}}`.
+And for g, it is harder to write as only part of :code:`op_2_()` is used when computing :code:`y`, but it can be written as :math:`v^T J_g = v^T \dfrac{\partial{y_1}}{\partial{y_0}} \dfrac{\partial{y_0}}{\partial{z_0}} \dfrac{\partial{z_0}}{\partial{x}}`.
+
+Computing not gradients
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In some cases, the user might need to compute something that is not the "true gradient" of its function.
+We identify few use cases here:
+
+- Ignore the gradients for some part of the computation as they could render the computed gradients unstable: gradients of square root near :math:`0`.
```

Comment:
nit: that's not the only case, right? if so, I would say "e.g." instead of ":", as in I would add "...gradients unstable, e.g. gradients of square root..."

same below